### ISSUE TITLE
Feature Request? Viewing []Bytes as a single Bytes

### ISSUE THREAD

**Issue by juliusl** [tokio-rs/bytes#729]
Say I have multiple Bytes references, and I want to view them as a single Bytes reference without allocating a BytesMut, how would I go about that? And if that is not currently possible, is that a feasible feature to add?

For example,

```rs
let a = Bytes::copy_from_slice([0, 1, 2]);
let b = Bytes::copy_from_slice([3, 4, 5]);

let ab: Bytes = Bytes::combined([a, b]);
```

This would allow for example, `ab.as_slice()[2..4]` which would create an iterator that would return `[2, 3]`. 

I'm sure I can solve this a different way, but just wanted to see if there's a canonical way of doing this inside of the bytes crate.

**Comment by seanmonstar** [comment]
Bytes always stores all of its data as a single contiguous slice. So, are you asking for a way to easily allocate a new buffer and copy all the bytes into it?

Or do you mean combine without allocation? In that case, perhaps you'd want to reach for an `impl Buf`, which doesn't promise to be a single contiguous slice. A simple one exists within the crate, `Chain`. But new ones can be implemented too.

If you mean a 3rd option of combining specifically in `Bytes`, but without an allocation, I don't see how that could be possible.

**Comment by juliusl** [comment]
@seanmonstar I meant combine without allocation, so I think Chain is what I'm looking for!

However, re: the 3rd option,

Say I have a big BytesMut that I am splitting off into smaller blocks. At some point I want to grab a range that might span over multiple blocks. 

Now, say I have those blocks, I can order them, so technically underneath the hood they are all part of the same contiguous chunk. In this case, would it be feasible to create a Bytes by using the pointer information from the first and last block?

**Comment by seanmonstar** [comment]
It might be possible, but not with the methods exposed today. And I wouldn't recommend depending on the internal details to do it externally.

**Comment by Darksonn** [comment]
The bytes crate has an [`unsplit`](https://docs.rs/bytes/latest/bytes/struct.BytesMut.html#method.unsplit) method.

I don't think we can otherwise support this, so I will close the feature request.

**Comment by juliusl** [comment]
@seanmonstar definitely, I wouldn't try to hack that together externally. 

But I was thinking about how with `try_reclaim` and `unsplit`, the original BytesMut is able to know that the reclaimed memory has no more references, so in the other direction would you be able to create a Bytes from the allocated memory space.

(@Darksonn I was hoping to just articulate the idea a bit more, so sorry to comment on a closed issue)

Basically, if I had an imaginary api like:

```rs
impl BytesMut {
  fn view_allocated(&self, offset: usize, len: usize) -> Option<Bytes>;
}
```

Would the owning `BytesMut` be able to know that all bytes along that range were frozen, such that returning a `Bytes` for that range would be safe.

**Comment by Darksonn** [comment]
No, `BytesMut` does not know which regions are frozen and which are part of a `BytesMut`.
